packages/SwingSet/src/kernel/vatManager/vat-warehouse.js

```diff
@@ -0,0 +1,257 @@
+// @ts-check
+import { assert } from '@agoric/assert';
+import { makeVatTranslators } from '../vatTranslator';
+
+/**
+ * @param { ReturnType<typeof import('../state/kernelKeeper').default> } kernelKeeper
+ * @param { ReturnType<typeof import('../loadVat').makeVatLoader> } vatLoader
+ * @param {{ sizeHint?: number }=} policyOptions
+ *
+ * @typedef { ReturnType<typeof import('@agoric/swing-store-simple').initSwingStore> } SwingStore
+ * @typedef {(syscall: VatSyscallObject) => ['error', string] | ['ok', null] | ['ok', Capdata]} VatSyscallHandler
+ * @typedef {{ body: string, slots: unknown[] }} Capdata
+ * @typedef { [unknown, ...unknown[]] } Tagged
+ * @typedef { { moduleFormat: string }} Bundle
+ */
+export function makeVatWarehouse(kernelKeeper, vatLoader, policyOptions) {
+  const { sizeHint = 20 } = policyOptions || {};
+
+  /**
+   * @typedef {{
+   *   manager: VatManager,
+   *   enablePipelining: boolean, //@@@put in DB. avoid spinning up a vat that isn't pipelined, in some cases.
+   * }} VatInfo
+   * @typedef { ReturnType<typeof import('../vatTranslator').makeVatTranslators> } VatTranslators
+   */
+  const ephemeral = {
+    /** @type {Map<string, VatInfo> } key is vatID */
+    vats: new Map(),
+  };
+
+  /** @type {Map<string, VatTranslators> } */
+  const xlate = new Map();
+  /** @param { string } vatID */
+  function findOrCreateTranslators(vatID) {
+    let translators = xlate.get(vatID);
+    if (!translators) {
+      // @@@ document pre-condition makeVatTranslators needs vatKeeper?
+      // or pass in the vatKeeper too;
+      // grep for getVatKeeper and avoid this pattern.
+      // remove ephemeral table from kernelKeeper@@@
+      translators = makeVatTranslators(vatID, kernelKeeper);
+      xlate.set(vatID, translators);
+    }
+    return translators;
+  }
+
+  /**
+   * @param {string} vatID
+   * @param {boolean} recreate
+   * @returns { Promise<VatInfo> }
+   */
+  async function ensureVatOnline(vatID, recreate) {
+    const info = ephemeral.vats.get(vatID);
+    if (info) return info;
+
+    // @@@getVatKeeper -> provideVatKeeper?
+    const vatKeeper =
+      kernelKeeper.getVatKeeper(vatID) || kernelKeeper.allocateVatKeeper(vatID);
+    const { source, options } = vatKeeper.getSourceAndOptions(); // @@@comms vat source repeated?
+
+    const translators = findOrCreateTranslators(vatID);
+
+    const createVat =
```

Comment:
personal style nit: I can tease out what the boolean short-circuits are doing, but I'd rather spend the brain cells elsewhere.. I'd write this with a pair of nested `if` statements.

```js
let createVat;
if (recreate) {
  if (kernelKeeper.getDynamicVats().includes(vatID)) {
    createVat = vatLoader.recreateDynamicVat;
  } else {
    createVat = vatLoader.recreateStaticVat;
  }
} else {
  createVat = vatLoader.createVatDynamically;
}
```

I do appreciate that your version is half the length of mine, though..

Reply:
yes, this is too clever by half.

for me, the issue isn't the length; it's that if statements aren't expressions, so `createVat` looks mutable and tsc doesn't infer a type.

nested `x ? y : z` would look more like those nested if statements, but we have eslint configured to prohibit that. maybe I should have just done a disable-this-line, i.e. be explicit that I'm stretching things a bit here.

Reply:
maybe make a tiny helper function? `function chooseLoader(recreate, kernelKeeper, vatID)` ?